newsroom/history_async.py

```diff
@@ -0,0 +1,178 @@
+
+from datetime import datetime
+from pydantic import BaseModel
+from typing import Optional, Dict, Annotated, List, Any, Union
+
+from newsroom import MONGO_PREFIX
+from newsroom.users.utils import get_user_or_abort
+from newsroom.utils import get_json_or_400
+from newsroom.core.resources.model import NewshubResourceModel
+from newsroom.core.resources.service import NewshubAsyncResourceService
+from newsroom.users.model import UserResourceModel
+
+from superdesk.core.module import Module
+from quart_babel import gettext
+from superdesk.utc import utcnow
+from superdesk.core.web import EndpointGroup, Response, Request
+from superdesk.flask import abort
+from superdesk.core import json
+from superdesk.core.resources.fields import ObjectId as ObjectIdField
+from eve.utils import ParsedRequest
+from superdesk.core.resources.validators import validate_data_relation_async
+from superdesk.core.resources import (
+    ResourceConfig,
+    MongoResourceConfig,
+    MongoIndexOptions,
+)
+
+
+class RouteParams(BaseModel):
+    item: Optional[Dict[str, Any]] = None
+    action: Optional[str] = None
+    section: Optional[str] = None
+
+
+class HistoryResourceModel(NewshubResourceModel):
+    action: Optional[str] = None
+    versioncreated: Optional[datetime] = None
+    user: Optional[Annotated[ObjectIdField, validate_data_relation_async("users")]] = None
+    company: Optional[Annotated[ObjectIdField, validate_data_relation_async("companies")]] = None
+    item: Optional[str] = None
+    version: Optional[str] = None
+    section: Optional[str] = None
+    extra_data: Optional[Dict] = None
+
+
+class HistoryService(NewshubAsyncResourceService[HistoryResourceModel]):
+    async def create(
+        self, docs: List[Dict[str, Any]], action: str, user: UserResourceModel, section: str = "wire", **kwargs: Any
+    ):
+        now = utcnow()
+
+        async def transform(item: Dict[str, Any]) -> Dict[str, Any]:
+            return {
+                "_id": ObjectId(),
+                "action": action,
+                "versioncreated": now,
+                "user": user.id,
+                "company": user.company,
+                "item": item["_id"],
+                "version": str(item.get("version", item.get("_current_version"))),
+                "section": section,
+            }
+
+        for doc in docs:
+            try:
+                await super().create([await transform(doc)])
+            except (werkzeug.exceptions.Conflict, pymongo.errors.BulkWriteError):
+                continue
+
+    async def create_history_record(
+        self, items: List[Dict[str, Any]], action: str, user: UserResourceModel, section: str
+    ):
+        await self.create(items, action, user, section)
+
+    async def query_items(self, query: Dict[str, Any]):
+        if query["from"] >= 1000:
+            # https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html#pagination
+            return abort(400)
+
+        req = ParsedRequest()
```

Comment:
This is from Eve. What you can do instead is using the `find` directly from the service. `return await self.find(query)`